src/Servers/Kestrel/Core/src/Internal/KestrelServerImpl.cs

```diff
@@ -247,6 +250,25 @@ async Task OnBind(ListenOptions options, CancellationToken onBindCancellationTok
 
             // Register the options with the event source so it can be logged (if necessary)
             KestrelEventSource.Log.AddServerOptions(Options);
+
+            static EndPoint ResolveEndPoint(EndPoint originalEndPoint, EndPoint currentEndPoint)
+            {
+                // From the original endpoint we want the address. This address could be a constant.
+                // For example, IPAddress.IPv4Any or IPAddress.IPv6Any. QUIC connection listener
+                // has logic that special cases these constants.
+                //
+                // However, from the current endpoint we want to get the port. If there is a port of 0
+                // then the first listener will resolve it to an actual port number. We want additional
+                // listeners to use that port number instead of each using 0 and resolving to different
+                // port numbers.
+                if (originalEndPoint is IPEndPoint originalIP && originalIP.Port == 0 &&
+                    currentEndPoint is IPEndPoint currentIP && currentIP.Port != 0)
+                {
+                    return new IPEndPoint(originalIP.Address, currentIP.Port);
+                }
+
+                return originalEndPoint;
```

Comment:
Let's continue returning `originalEndPoint` for non-`IPEndPoint`s. I don't know of a specific scenario it fixes, but custom transports could be changing non-`IPEndPoint`s on bind.